Add tsan_release edge on task creation

yln opened the pull request:
We need to update the modeling of Swift Concurrency operations for TSan to avoid false positives.

For the original TSan annotations, I documented my understanding of the operations and reasoning for where to put the TSan edges in the commit messages: [initial commit](https://github.com/apple/swift/commit/a3ccdd29d8db2702499e323da93859d342b02c7c), [task groups](https://github.com/apple/swift/commit/29107180e50576b87f8c59da6ed9b7c6b71210f7).
Since then the Swift Concurrency Runtime has evolved and we need to refine our TSan annotations. Relevant commits: [1](https://github.com/apple/swift/commit/b9ee090605029e830ee9f82d18e57b854dcda17d), [2](https://github.com/apple/swift/commit/1a9dc145a6a65913459d4ec0ad7ca018b6e17cd0), ...

**Fix: Add tsan_release edge on task creation**

Without this, we are getting false data races when a task is created and immediately scheduled on a different thread.

False positive for `Sanitizers/tsan/actor_counters.swift` test:
```
WARNING: ThreadSanitizer: data race (pid=81452)
  Read of size 8 at 0x7b2000000560 by thread T5:
    #0 Counter.next() <null>:2 (a.out:x86_64+0x1000047f8)
    #1 (1) suspend resume partial function for worker(identity:counters:numIterations:) <null>:2 (a.out:x86_64+0x100005961)
    #2 swift::runJobInEstablishedExecutorContext(swift::Job*) <null>:2 (libswift_Concurrency.dylib:x86_64+0x280ef)

  Previous write of size 8 at 0x7b2000000560 by main thread:
    #0 Counter.init(maxCount:) <null>:2 (a.out:x86_64+0x1000046af)
    #1 Counter.__allocating_init(maxCount:) <null>:2 (a.out:x86_64+0x100004619)
    #2 runTest(numCounters:numWorkers:numIterations:) <null>:2 (a.out:x86_64+0x100006d2e)
    #3 swift::runJobInEstablishedExecutorContext(swift::Job*) <null>:2 (libswift_Concurrency.dylib:x86_64+0x280ef)
    #4 main <null>:2 (a.out:x86_64+0x10000a175)
```

New edge with this change:
```
[4357150208] allocate task 0x7b3800000000, parent = 0x0
[4357150208] creating task 0x7b3800000000 with parent 0x0
[4357150208] tsan_release on 0x7b3800000000    <<< new release edge
[139088221442048] tsan_acquire on 0x7b3800000000
[139088221442048] trying to switch from executor 0x0 to 0x7ff85e2d9a00
[139088221442048] switch failed, task 0x7b3800000000 enqueued on executor 0x7ff85e2d9a00
[139088221442048] enqueue job 0x7b3800000000 on executor 0x7ff85e2d9a00
[139088221442048] tsan_release on 0x7b3800000000
[139088221442048] tsan_release on 0x7b3800000000
[4357150208] tsan_acquire on 0x7b3800000000
counters: 1, workers: 1, iterations: 1
[4357150208] allocate task 0x7b3c00000000, parent = 0x0
[4357150208] creating task 0x7b3c00000000 with parent 0x0
[4357150208] tsan_release on 0x7b3c00000000    <<< new release edge
[139088221442048] tsan_acquire on 0x7b3c00000000
[4357150208] task 0x7b3800000000 waiting on task 0x7b3c00000000, going to sleep
[4357150208] tsan_release on 0x7b3800000000
[4357150208] tsan_release on 0x7b3800000000
[139088221442048] getting current executor 0x0
[139088221442048] tsan_release on 0x7b3c00000000
...
```

Note: TSan annotations should surround "synchronization operations" as follows:
```
 --- T1 ---   --- T2 ---

tsan_release(addr)
  sync()
    ...    -->   ...
               sync()
             tsan_acquire(addr)
```

rdar://78932849

Commit:
Synchronize both versions of actor_counters.swift test

Commit:
Synchronize on Job address

Make sure to synchronize on Job address (AsyncTasks are Jobs, but not
all Jobs are AsyncTasks).

Commit:
Add fprintf debug output for TSan acquire/release

Commit:
Add tsan_release edge on task creation

without this, we are getting false data races between when a task
is created and immediately scheduled on a different thread.

False positive for `Sanitizers/tsan/actor_counters.swift` test:
```
WARNING: ThreadSanitizer: data race (pid=81452)
  Read of size 8 at 0x7b2000000560 by thread T5:
    #0 Counter.next() <null>:2 (a.out:x86_64+0x1000047f8)
    #1 (1) suspend resume partial function for worker(identity:counters:numIterations:) <null>:2 (a.out:x86_64+0x100005961)
    #2 swift::runJobInEstablishedExecutorContext(swift::Job*) <null>:2 (libswift_Concurrency.dylib:x86_64+0x280ef)

  Previous write of size 8 at 0x7b2000000560 by main thread:
    #0 Counter.init(maxCount:) <null>:2 (a.out:x86_64+0x1000046af)
    #1 Counter.__allocating_init(maxCount:) <null>:2 (a.out:x86_64+0x100004619)
    #2 runTest(numCounters:numWorkers:numIterations:) <null>:2 (a.out:x86_64+0x100006d2e)
    #3 swift::runJobInEstablishedExecutorContext(swift::Job*) <null>:2 (libswift_Concurrency.dylib:x86_64+0x280ef)
    #4 main <null>:2 (a.out:x86_64+0x10000a175)
```

New edge with this change:
```
[4357150208] allocate task 0x7b3800000000, parent = 0x0
[4357150208] creating task 0x7b3800000000 with parent 0x0
[4357150208] tsan_release on 0x7b3800000000    <<< new release edge
[139088221442048] tsan_acquire on 0x7b3800000000
[139088221442048] trying to switch from executor 0x0 to 0x7ff85e2d9a00
[139088221442048] switch failed, task 0x7b3800000000 enqueued on executor 0x7ff85e2d9a00
[139088221442048] enqueue job 0x7b3800000000 on executor 0x7ff85e2d9a00
[139088221442048] tsan_release on 0x7b3800000000
[139088221442048] tsan_release on 0x7b3800000000
[4357150208] tsan_acquire on 0x7b3800000000
counters: 1, workers: 1, iterations: 1
[4357150208] allocate task 0x7b3c00000000, parent = 0x0
[4357150208] creating task 0x7b3c00000000 with parent 0x0
[4357150208] tsan_release on 0x7b3c00000000    <<< new release edge
[139088221442048] tsan_acquire on 0x7b3c00000000
[4357150208] task 0x7b3800000000 waiting on task 0x7b3c00000000, going to sleep
[4357150208] tsan_release on 0x7b3800000000
[4357150208] tsan_release on 0x7b3800000000
[139088221442048] getting current executor 0x0
[139088221442048] tsan_release on 0x7b3c00000000
...
```

rdar://78932849

yln committed:
Merge branch 'main' into tsan-actor_counters.swift

yln commented:
@swift-ci Please test

yln committed:
Add static_cast<Job *>()

yln commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -642,6 +642,8 @@ static AsyncTaskAndContext swift_task_create_commonImpl(
     swift_task_enqueue(task, executor);
   }
+  
+  _swift_tsan_release(static_cast<Job *>(task));
```
Actually, the false positive went away again after https://github.com/apple/swift/commit/3727db2fe2f3bd9c1f4b75ba3343c5445de53740 because the call to `swift_task_enqueue()` adds the required release edge.

@DougGregor 
Are all tasks that are created here, i.e., also the ones that are not immediately enqueued above (`taskCreateFlags.enqueueJob() == false`) always guaranteed to be scheduled via `swift_task_enqueue()` before they start executing on a different thread?  If so, then we don't need this anymore.

DougGregor replied:
They will be scheduled via `swift_task_enqueue` at some point, yes.

yln replied:
Do you know what changed here?  Previously, not all tasks were going through the `swift_task_enqueue()` funnel, which produced the false positive we were seeing.

Do you recommend adding this (superfluous) edge here to protect us from future changes, or relying on "all tasks go through `swift_task_enqueue ()` even if not immediately scheduled"?

yln replied:
Doug pointed out that some places call `swift_task_enqueueGlobal()` directly.

The most common (and currently covered) case seems to be `swift_task_enqueue()` calling `swift_task_enqueueGlobal()`, but if the latter (more specialized) version is getting called “by itself”, we should add an edge there (and actually not at task creation).

yln replied:
I moved the TSan release edge from `swift_task_create_commonImpl()` to `swift_task_enqueueGlobalImpl()`.  Task creation itself is not an event that needs synchronization, but rather that task creation "happens before" execution of that task on another thread.

This edge is usually added when the task is scheduled via `swift_task_enqueue()` (which then usually calls `swift_task_enqueueGlobal()`).  However, not all task scheduling goes through the `swift_task_enqueue()` funnel: some places call the more specific `swift_task_enqueueGlobal()` directly.  So let's annotate this function (duplicate edges aren't harmful) to ensure we cover all schedule events, including newly-created tasks (our original problem).

yln commented:
@swift-ci Please test

yln commented:
> **Build failed**
> [Swift Test OS X Platform](https://ci.swift.org/job/swift-PR-macos/28018/)
> Git Sha - [3c821de](https://github.com/apple/swift/commit/3c821deb5df8ba26e56aa33a0c13ab0a38803f2e)

Unrelated tests:
```
22:37:14 Failed Tests (2):
22:37:14   Swift(macosx-x86_64) :: Interpreter/SDK/objc_bridge_cast.swift
22:37:14   Swift(macosx-x86_64) :: Interpreter/SDK/objc_dealloc.swift
```

jckarter reviewed (approved):
Makes sense.

Commit:
Move TSan release edge to swift_task_enqueueGlobal()

Move the TSan release edge from `swift_task_create_commonImpl()` to
`swift_task_enqueueGlobalImpl()`.  Task creation itself is not an event
that needs synchronization, but rather that task creation "happens
before" execution of that task on another thread.

This edge is usually added when the task is scheduled via
`swift_task_enqueue()` (which then usually calls
`swift_task_enqueueGlobal()`).  However, not all task scheduling goes
through the `swift_task_enqueue()` funnel as some places call the more
specific `swift_task_enqueueGlobal()` directly.  So let's annotate this
function (duplicate edges aren't harmful) to ensure we cover all
schedule events, including newly-created tasks (our original problem
here).

rdar://78932849

yln commented:
@swift-ci Please test

ktoso reviewed (approved):
Makes sense, thinks!
